src/com/google/cloud/datastore/types.js

```diff
@@ -0,0 +1,301 @@
+      var values = new Array(a.length);
+      for ( var i = 0; i < a.length; i++) {
+        values[i] = com.google.cloud.datastore.toDatastoreValue(a[i]);
+      }
+      return { arrayValue: { values: values } };
+    },
+    function fromDatastoreValue(v) {
+      var values = v.arrayValue;
+      var arr = new Array(values.length);
+      for ( var i = 0; i < values.length; i++ ) {
+        arr[i] = com.google.cloud.datastore.fromDatastoreValue(values[i]);
+      }
+      return arr;
+    }
+  ]
+});
+
+foam.LIB({
+  name: 'foam.Date',
+
+  methods: [
+    function toDatastoreValue(d) {
+      return { timestampValue: d.toISOString() };
+    },
+    function fromDatastoreValue(v) {
+      return new Date(Date.parse(v.timestampValue));
+    }
+  ]
+});
+
+foam.LIB({
+  name: 'foam.core.FObject',
+
+  methods: [
+    function toDatastoreValue(o) { return o.toDatastoreValue(); },
+    function fromDatastoreValue(v) {
+      return this.fromDatastoreEntity(v.entityValue);
+    },
+    function fromDatastoreEntity(entity) {
+      var keys = entity.key.path;
+      var key = keys[keys.length - 1];
+      var cls = foam.lookup(key.kind);
+      var id = key.name;
+
+      var o = cls.create();
+
+      if ( cls.ids && cls.ids.length > 1 ) {
+        throw new Error('Not implemented: Deserialization of Cloud Datastore ' +
+            'multi-part ids');
+      }
+
+      var idProp = cls.ids && cls.ids.length === 1 ?
+          cls.getAxiomByName(cls.ids[0]) :
+          cls.getAxiomByName('id');
+
+      if ( idProp ) o[idProp.name] = id;
+
+      var props = entity.properties;
+      for ( var name in props ) {
+        if ( props.hasOwnProperty(name) ) {
+          o[name] = com.google.cloud.datastore.fromDatastoreValue(props[name]);
+        }
+      }
+
+      return o;
+    },
+    function getOwnClassDatastoreKind() {
+      return this.id;
+    },
+    function getClassDatastoreKind() {
+      return { name: this.getOwnClassDatastoreKind() };
+    }
+  ]
+});
+
+//
+// Provide base to/from datastore value operations.
+//
+
+foam.LIB({
+  name: 'com.google.cloud.datastore',
+
+  methods: [
+    function toDatastoreValue(o) {
+      var t = foam.typeOf(o);
+      if ( ! ( t && t.toDatastoreValue ) ) {
+        throw new Error('Attempt to compute datastore value from ' +
+            'incompatible type');
+      }
+
+      return t.toDatastoreValue(o);
+    },
+    function fromDatastoreValue(v) {
+      return this.typeOfDatastoreValue(v).fromDatastoreValue(v);
+    },
+    function fromDatastoreEntity(v) {
+      return foam.core.FObject.fromDatastoreEntity(v);
+    },
+    (function() {
+      var typeMap = {
+        nullValue: foam.Null,
+        booleanValue: foam.Boolean,
+        integerValue: foam.Number,
+        doubleValue: foam.Number,
+        timestampValue: foam.Date,
+        stringValue: foam.String,
+        entityValue: foam.core.FObject
+      };
+      return function typeOfDatastoreValue(v) {
```

Comment:
That's how https://cloud.google.com/datastore/docs/reference/rest/v1/projects/runQuery#value is modeled. I know: *eye-roll*. Documentation added; done.